onnxruntime/contrib_ops/cpu/quantization/qlinear_softmax.cc

```diff
@@ -0,0 +1,358 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "contrib_ops/cpu/quantization/qlinear_softmax.h"
+
+#include <cstdint>
+#include <type_traits>
+#include <utility>
+
+#include "core/common/common.h"
+#include "core/framework/tensorprotoutils.h"
+#include "core/providers/common.h"
+#include "core/providers/cpu/tensor/transpose.h"
+
+#include "core/mlas/inc/mlas.h"
+#include "core/platform/threadpool.h"
+#include "gsl/gsl-lite.hpp"
+
+namespace onnxruntime {
+namespace contrib {
+
+constexpr int OPSET13 = 13;
+
+namespace {
+
+void QlinearBuildLookupTableUint32(gsl::span<uint32_t> table,
+                                   const float x_scale,
+                                   size_t reduce_len, bool is_signed) {
+  const double qscale =
+      fmin(static_cast<double>(UINT32_MAX) / static_cast<double>(reduce_len), static_cast<double>(0x7fffff));
+  for (int32_t i = 0; i < 256; i++) {
+    double scaled_exp_xi = qscale * exp(static_cast<double>(i - 255) * static_cast<double>(x_scale));
+    // we can't get the real max value of input tensor here, so we just assume 255.
+    // in the function of `QlinearSoftmaxCPU`,
+    // all numbers will have a shift (255-max_value) if its max value is not 255
+    //
+    // if is_signed index = [1 2 3 ......126 127 -128 -127 ..... -3 -2 -1]
+    // else [0 1 2 3 4 ..... 256]
+    uint8_t index = static_cast<uint8_t>(is_signed ? i - 128 : i);
+    table[index] = static_cast<uint32_t>(lrint(scaled_exp_xi));
+  }
+}
+
+void BuildLookupTableIfFixed(const OpKernelInfo& info, std::vector<uint32_t>& fixed_lookup_table,
+                             size_t reduce_len, bool is_signed) {
+  const Tensor* tensor_x_scale = nullptr;
+
+  bool get_x_scale = info.TryGetConstantInput(1, &tensor_x_scale);
+  ORT_ENFORCE(tensor_x_scale == nullptr || IsScalarOr1ElementVector(tensor_x_scale),
+              "QlinearBuildLookupTable : input X_scale must be a scalar or 1D tensor of size 1");
+  bool is_fixed_parameters = get_x_scale;
+
+  if (is_fixed_parameters) {
+    fixed_lookup_table.resize(256);
+    const float X_scale = *(tensor_x_scale->Data<float>());
+    QlinearBuildLookupTableUint32(fixed_lookup_table, X_scale, reduce_len, is_signed);
+  }
+}
+}  // namespace
+
+
+QLinearSoftmax::QLinearSoftmax(const OpKernelInfo& info)
+    : OpKernel(info) {
+  const auto& node = info.node();
+  auto input_defs = node.InputDefs();
+  auto input_type = input_defs[0]->TypeAsProto()->tensor_type().elem_type();
+  is_signed_ = (input_type == ONNX_NAMESPACE::TensorProto_DataType_INT8);
+  const auto* x_shape = input_defs[0]->Shape();
+  if (x_shape == nullptr || x_shape->dim_size() == 0) {
+    return;
+  }
```

Comment:
This is bad as it's a silent early exit that leaves axis_ unitialized. If that happened it would be really hard to debug. 

There should always be a shape by this point and the rank (dim_size()) should be known, so ORT_ENFORCE would be more appropriate.

Reply:
Enforce checked.